Client/Migration/BraveCoreMigrator.swift

```diff
@@ -128,36 +128,53 @@ public class BraveCoreMigrator {
       return
     }
 
+    let group = DispatchGroup()
+    
+    group.enter()
+    
     // Step 1:  Check If bookmarks are migrated / migrate
     migrateBookmarkModels { [unowned self] success in
       guard success else {
         self.migrationObserver = .failed
         completion?(.failedBookmarksMigration)
-
+        
         return
       }
+      
+      group.leave()
+    }
+    
+    group.enter()
+    
+    // Step 2: Check If history is migrated / migrate
+    migrateHistoryModels { [unowned self] success in
+      guard success else {
+        self.migrationObserver = .failed
+        completion?(.failedHistoryMigration)
+        
+        return
+      }
+      
+      group.leave()
+    }
+    
+    group.enter()
+    
+    // Step 3: Check If passwords are migrate / migrate
+    migratePasswordForms { [unowned self] success in
+      guard success else {
+        self.migrationObserver = .failed
+        completion?(.failedPasswordMigration)
```

Comment:
You're also risking multiple callbacks to completion with all of these because the `return` below is scoped to this completion block. 

You should instead call it only once in the `group.notify` using a failure reason set elsewhere

Reply:
Fix at https://github.com/brave/brave-ios/pull/5641/commits/e745114a3e090226f302b99b880622defa67ea46